Adding dynamic metric collection for SNO (#3453)
open-cluster-management/rhacm-docs: clusters/single_node_openshift.adoc

[#single-node-openshift]
= Creating a single node OpenShift cluster (Technology Preview)

> //should tech preview be removed from the title 

Reviewer: @cadawson

. Complete the remaining steps of the cluster creation to create your SNO cluster.
+
*Important:* When you create an {ocp-short} SNO cluster on GCP, you must configure more CPU and memory resources than for a standard GCP cluster. The minimum requirements for a SNO cluster on GCP is 8 CPUs and 32 GB of memory. The standard requirements of 4 CPUs and 16 GB of memory is not stable for this type of cluster. 

[#dynamic-metrics-for-sno]
== Dynamic metrics for SNO clusters

> Dynamic metric collection for SNO clusters is supported with a group of collection rules (`collection_rules`). Conditions for the collection rules are written as a PromQL expression. By default, the following conditions are defined: `HighCPUUsage` and `HighMemoryUsage`. When eiher of the conditions exceed 70 percent continuously, the observability component collects data. Observability automatically collects pod and container metrics for SNO clusters. 

Author: We might want to add an example of the yaml file for this parameter section

Reviewer 2: @dockerymick I think we need to add a introduction to what dynamic metrics collection means. 

Dynamic metrics collection refers to the ability to collect metrics on automatically based on certain condition.  Starting with ACM 2.5, By default, a SNO cluster does not collect pod and container resource metrics. Once a SNO cluster reaches a level of resource consumption, these granular metrics are collected dynamically. When the cluster resource consumption is consistently less than the threshold for a period of time, collection of the granular metrics stops.

The metrics are collected dynamically based on condition on the managed cluster specified by a `collection rule`. Because these metrics are collected dynamically, the following ACM grafana dashboards will not have any data to show. When a collection rule is activated and the corresponding metrics are collected, these panels will show data for the duration of the time the collection rule is triggered. 
- Kubernetes / Compute Resources / Namespace (Pods)
- Kubernetes / Compute Resources / Namespace (Workloads)
- Kubernetes / Compute Resources / Nodes (Pods)
- Kubernetes / Compute Resources / Pod
- Kubernetes / Compute Resources / Workload

A `collection rule` defines the following:
-  a set of metrics to collect dynamically
-  a trigger condition written as a PromQL expression
-  a time interval for which the collect rule must evaluate to `true`
- a match expression that allows to select clusters where the collect rule must be evaluated

Collection rules are evaluated continuously on managed clusters every 30 seconds or the metrics collection interval specified, whichever is less. Once the collection rule condition persists for the duration specified by the `for` attribute, the collection rule triggers and the metrics specified by the rule are automatically collected on the managed cluster. Metrics collection stops automatically after the collection rule condition no longer exists on the managed cluster at least 15 minutes after it is triggered. 

The `Collection rules` are grouped together as a named group under `collect_rules`  for administrative convenience so they can be enabled or disabled as a group. 

ACM 2.5  release defines a collection rule group `SNOResourceUsage`  with two default collection rules `HighCPUUsage` and `HighMemoryUsage` (see the yaml definition below). `HighCPUUsage` collection rule triggers when the node CPU usage exceeds 70%. `HighMemoryUsage` collection rule is triggered if the overall memory utilization of the SNO cluster exceeds 70% of the available node memory. Currently, these thresholds are fixed, and cannot be changed for the current release. When a collection rule triggers for more than the interval specified by `for` attribute interval, the system automatically starts collecting the metrics specified under the `dynamic_metrics` section. 

Please refer to the list of dynamic metrics that would be collected under each collection rule specified in the yaml below. 

```
collect_rules:
  - group: SNOResourceUsage
    annotations:
      description: >
        By default, a SNO cluster does not collect pod and container resource metrics. Once a SNO cluster 
        reaches a level of resource consumption, these granular metrics are collected dynamically. 
        When the cluster resource consumption is consistently less than the threshold for a period of time, 
        collection of the granular metrics stops.
    selector:
      matchExpressions:
        - key: clusterType
          operator: In
          values: ["SNO"]
    rules:
    - collect: SNOHighCPUUsage
      annotations:
        description: >
          Collects the dynamic metrics specified if the cluster cpu usage is constantly more than 70% for 2 minutes
      expr: (1 - avg(rate(node_cpu_seconds_total{mode=\"idle\"}[5m]))) * 100 > 70
      for: 2m
      dynamic_metrics:
        names:
          - container_cpu_cfs_periods_total
          - container_cpu_cfs_throttled_periods_total
          - kube_pod_container_resource_limits 
          - kube_pod_container_resource_requests   
          - namespace_workload_pod:kube_pod_owner:relabel 
          - node_namespace_pod_container:container_cpu_usage_seconds_total:sum_irate 
          - node_namespace_pod_container:container_cpu_usage_seconds_total:sum_rate 
    - collect: SNOHighMemoryUsage
      annotations:
        description: >
          Collects the dynamic metrics specified if the cluster memory usage is constantly more than 70% for 2 minutes
      expr: (1 - sum(:node_memory_MemAvailable_bytes:sum) / sum(kube_node_status_allocatable{resource=\"memory\"})) * 100 > 70
      for: 2m
      dynamic_metrics:
        names:
          - kube_pod_container_resource_limits 
          - kube_pod_container_resource_requests 
          - namespace_workload_pod:kube_pod_owner:relabel
        matches:
          - __name__="container_memory_cache",container!=""
          - __name__="container_memory_rss",container!=""
          - __name__="container_memory_swap",container!=""
          - __name__="container_memory_working_set_bytes",container!=""
```

A `collection rule group` can be disabled in `custom-allowlist` as shown below. When a `collection rule group` is disabled, metrics collection reverts to the old behavior; these metrics will be collected at regularly specified interval as in previous releases.

```
collect_rules:
  - group: -SNOResourceUsage
```

Author: @subbarao-meduri Thanks for your comments and patience. I have updated the PR. When you have an opportunity, will you re-review the PR please?

. Complete the remaining steps of the cluster creation to create your SNO cluster.
+
*Important:* When you create an {ocp-short} SNO cluster on GCP, you must configure more CPU and memory resources than for a standard GCP cluster. The minimum requirements for a SNO cluster on GCP is 8 CPUs and 32 GB of memory. The standard requirements of 4 CPUs and 16 GB of memory is not stable for this type of cluster. 

[#dynamic-metrics-for-sno]
== Dynamic metrics for SNO clusters

> Dynamic metrics collection supports automatic metric collection based on certain conditions. By default, a SNO cluster does not collect pod and container resource metrics. Once a SNO cluster reaches a specific level of resource consumption, the defined granular metrics are collected dynnamically. When the cluster resource consumption is consistently less than the threshold for a period of time, granular metric collection stops.

Reviewer 3: ```suggestion
Dynamic metrics collection supports automatic metric collection based on certain conditions. By default, a SNO cluster does not collect pod and container resource metrics. Once a SNO cluster reaches a specific level of resource consumption, the defined granular metrics are collected dynamically. When the cluster resource consumption is consistently less than the threshold for a period of time, granular metric collection stops.
```

Reviewer 3: Dynamically is misspelled.

Author: Great catch!

The metrics are collected dynamically based on the conditions on the managed cluster specified by a collection rule. Because these metrics are collected dynamically, the following {product-title-short} Grafana dashboards do not display any data. When a collection rule is activated and the corresponding metrics are collected, the following panels display data for the duration of the time that the collection rule is initiated:

* Kubernetes / Compute Resources / Namespace (Pods)
* Kubernetes / Compute Resources / Namespace (Workloads)
* Kubernetes / Compute Resources / Nodes (Pods)
* Kubernetes / Compute Resources / Pod
* Kubernetes / Compute Resources / Workload

A collection rule includes the following conditions:

* A set of metrics to collect dynamically.
* Conditions written as a PromQL expression.
* A time interval for the collection, which must be set to `true`.

> * A match expressiion to select clusters where the collect rule must be evaluated.

Reviewer 3: ```suggestion
* A match expression to select clusters where the collect rule must be evaluated.
```

Reviewer 3: Expression is misspelled.

Author: Thanks again!

//Is it important to say "whichever is less"
By default, collection rules are evaluated continuously on managed clusters every 30 seconds, or at the time interval that is specified. Once the collection rule condition persists for the duration specified by the `for` attribute, the collection rule starts and the metrics specified by the rule are automatically collected on the managed cluster. Metrics collection stops automatically after the collection rule condition no longer exists on the managed cluster, at least 15 minutes after it starts.

> The collection rules are grouped together as a parameter section named `collection_rules`, where it can be enabled or disabled as a group. {product-title-short} isntallation includes the collection rule ggroup, `SNOResourceUsage` with two default collection rules: `HighCPUUsage` and `HighMemoryUsage`. The `HighCPUUsage` collection rule begins when the node CPU usage exceeds 70%. The `HighMemoryUsage` collection rule begins if the overall memory utilization of the SNO cluster exceeds 70% of the availale node memory. Currently, the previously mentioned thresholds are fixed and cannot be changed. When a collection rule begins for more than the interval specified by the `for` attribute, the system automatically starts collecting the metrics that are specified in the `dynamic_metrics` section.

Reviewer 3: ```suggestion
The collection rules are grouped together as a parameter section named `collection_rules`, where it can be enabled or disabled as a group. {product-title-short} installation includes the collection rule group, `SNOResourceUsage` with two default collection rules: `HighCPUUsage` and `HighMemoryUsage`. The `HighCPUUsage` collection rule begins when the node CPU usage exceeds 70%. The `HighMemoryUsage` collection rule begins if the overall memory utilization of the SNO cluster exceeds 70% of the available node memory. Currently, the previously mentioned thresholds are fixed and cannot be changed. When a collection rule begins for more than the interval specified by the `for` attribute, the system automatically starts collecting the metrics that are specified in the `dynamic_metrics` section.
```

Reviewer 3: There were a few more misspelled words in this section.

Author: Thanks Disaiah, I have some janky keys on my laptop. I wasn't using my wireless keyboard :P

> View the list of dynamic metrics that from the `collection_rule` section, in the following YAML file:

Reviewer 3: Should we change `collection_rule` to `collect_rules`? That way it would be consistent with the yaml key property.

Author: Sounds good, will add the change now

Author: updated

          - container_cpu_cfs_throttled_periods_total
          - kube_pod_container_resource_limits 
          - kube_pod_container_resource_requests   
          - namespace_workload_pod:kube_pod_owner:relabel 
          - node_namespace_pod_container:container_cpu_usage_seconds_total:sum_irate 
          - node_namespace_pod_container:container_cpu_usage_seconds_total:sum_rate 
    - collect: SNOHighMemoryUsage
      annotations:
        description: >
          Collects the dynamic metrics specified if the cluster memory usage is constantly more than 70% for 2 minutes
      expr: (1 - sum(:node_memory_MemAvailable_bytes:sum) / sum(kube_node_status_allocatable{resource=\"memory\"})) * 100 > 70
      for: 2m
      dynamic_metrics:
        names:
          - kube_pod_container_resource_limits 
          - kube_pod_container_resource_requests 
          - namespace_workload_pod:kube_pod_owner:relabel
        matches:
          - __name__="container_memory_cache",container!=""
          - __name__="container_memory_rss",container!=""
          - __name__="container_memory_swap",container!=""
          - __name__="container_memory_working_set_bytes",container!=""
----

> A `collection_rule.group` can be disabled in the `custom-allowlist` as shown in the following example. When a `collection_rule.group` is disabled, metrics collection reverts to the previous behavior. These metrics are collected at regularly, specified intervals:

Reviewer 3: Similarly here, should we change the `collection_rule.group` to `collect_rules.group` to help remain consistent throughout the doc?

Author: updated

The metrics are collected dynamically based on the conditions on the managed cluster specified by a collection rule. Because these metrics are collected dynamically, the following {product-title-short} Grafana dashboards do not display any data. When a collection rule is activated and the corresponding metrics are collected, the following panels display data for the duration of the time that the collection rule is initiated:

* Kubernetes / Compute Resources / Namespace (Pods)
* Kubernetes / Compute Resources / Namespace (Workloads)
* Kubernetes / Compute Resources / Nodes (Pods)
* Kubernetes / Compute Resources / Pod
* Kubernetes / Compute Resources / Workload

A collection rule includes the following conditions:

* A set of metrics to collect dynamically.
* Conditions written as a PromQL expression.
* A time interval for the collection, which must be set to `true`.
* A match expression to select clusters where the collect rule must be evaluated.

> //Is it important to say "whichever is less"

Reviewer 2: Yes, it is important to say "whichever is less" as  lower of the two values takes precedence. For example, the default collection interval is 300 seconds, but the collect rules are evaluated every 30 seconds. The interval can be set as low as 15 seconds, in which case, the collect rules also execute once every 15 seconds. Hope this helps.

Author: Update on line 85
